src/v/datalake/coordinator/iceberg_file_committer.cc

```diff
@@ -259,6 +260,132 @@ ss::future<checked<std::nullopt_t, file_committer::errc>> commit_icb_files(
     co_return std::nullopt;
 }
 
+/// A table_committer accumulates files to commit to an Iceberg table and
+/// manages the commit process.
+///
+/// A file_committer will usually create one or more table_committer-s per
+/// topic, and will call process_pending_entry for each pending entry in the
+/// topic's state. Once all pending entries have been processed, the
+/// file_committer must call commit to commit the files to the Iceberg table.
+class table_committer {
+private:
+    explicit table_committer(
+      iceberg::table_identifier table_id,
+      iceberg::table_metadata&& table,
+      std::optional<model::offset> table_commit_offset)
+      : table_id_(std::move(table_id))
+      , table_(std::move(table))
+      , table_commit_offset_(table_commit_offset) {}
+
+public:
+    static checked<table_committer, file_committer::errc> create(
+      iceberg::table_identifier table_id, iceberg::table_metadata&& table) {
+        auto meta_res = get_iceberg_committed_offset(table);
+        if (meta_res.has_error()) {
+            vlog(
+              datalake_log.warn,
+              "Error getting snapshot property '{}' for table {}: {}",
+              commit_meta_prop,
+              table_id,
+              meta_res.error());
+            return file_committer::errc::failed;
+        }
+
+        return table_committer(
+          std::move(table_id), std::move(table), meta_res.value());
+    }
+
+public:
+    checked<std::nullopt_t, file_committer::errc> process_pending_entry(
+      const model::topic& topic,
+      model::revision_id topic_revision,
+      const iceberg::manifest_io& io,
+      const model::offset added_pending_at,
+      const chunked_vector<data_file>& files) {
+        if (skip_entry(added_pending_at)) {
+            // This entry was committed to the Iceberg table already.
+            // Intentionally collect the pending commit above so we can
+            // replicate the fact that it was committed previously, but
+            // don't construct a data_file to send to Iceberg as it is
+            // already committed.
+            vlog(
+              datalake_log.debug,
+              "Skipping entry for topic {} revision {} added at "
+              "coordinator offset {} because table {} has data including "
+              "coordinator offset {}",
+              topic,
+              topic_revision,
+              added_pending_at,
+              table_id_,
+              table_commit_offset_);
+        } else {
+            auto res = collect_icb_files(
```

Comment:
I'm not against keeping them, but I'm curious did the previous refactorings actually help reuse this code if we're pulling out a reusable table_committer abstraction?

Reply:
It took a few iterations to arrive at the table committer abstractions. :)

Had to stop somewhere otherwise there is much more that can be improved. Given that I need to address the feedback here I'll try to do a couple more iterations tidying this code then.